File: src/v8/experts/pandf_breakout.py
"""Point-and-Figure boxed-price breakout behavior family (`pandf_breakout`).

Hypothesis (mechanism `boxed_price_breakout`): a boxed-price representation of
the close series (columns of X rising / O falling, Ch15.1 p599) makes double
and triple top/bottom breakouts mechanically detectable — a new X above the
prior column's top is a buy signal, a new O below the prior column's bottom is
a sell signal (Ch15.2 p600-617). Targets use the book's vertical count:
`lowestBoxPrice + (breakoutColBoxes * boxSize * reversalSize)` (Ch15.3
p620-623). Stops sit below the lowest X in the breakout column (bullish) /
above the highest O in the breakout column (bearish) (Ch15.2 p617).

The box filter is DECLARED and LOCKED (orchestrator directive): box = 1.0 *
ATR at detection, reversal = 3 boxes. The P&F transform is computed INSIDE the
expert from the state's `history` close window (no marketstate box/reversal
representation exists — V8_LOGIC_GAP H07).

Documented deviations: (1) the catapult signal requires a two-stage partial
entry (position scaling), which is O-013-gated — not implemented (CRITIC 4.2);
(2) the ascending/descending/spread-triple signals are angle classifications
of the column ladder with no numeric book spec — not implemented; (3) box-grid
alignment is anchored to the column structure rather than an absolute price
grid (a volatility-based box has no natural absolute grid).

Variants (all frozen; D-044 lists every implemented variant):
  a  double-top box breakout (LONG): the current X column's top exceeds the
     prior X column's top.
  b  double-bottom box breakout (SHORT): the current O column's bottom is
     below the prior O column's bottom.
  c  triple-top box breakout (LONG): the current X column's top exceeds both
     prior X column tops.
  d  triple-bottom box breakout (SHORT): the current O column's bottom is
     below both prior O column bottoms.
"""
from __future__ import annotations

from ..schema import MarketState, CandidateDraft, ExpertEvaluation
from .base import Expert

# Declared, LOCKED box filter (orchestrator directive): box = 1.0 * ATR at
# detection, reversal = 3 boxes.
BOX_ATR_K = 1.0
REVERSAL_BOXES = 3
# Minimum history for a meaningful column structure.
MIN_HISTORY_BARS = 20


def _columns(closes: list, box: float, reversal: int):
    """Deterministic close-based P&F columns.

    Returns a list of (direction, start_idx, levels):
      direction: +1 = X column (rising), -1 = O column (falling)
      start_idx: index of the first close that contributed to the column
      levels: box prices in the column, ordered from the column's origin to
      its extreme (rising for X, falling for O).
    A bar that stays inside the current column's box range is absorbed
    (contributes nothing); a move of >= `reversal` boxes in the opposite
    direction starts a new column one box beyond the extreme.
    """
    cols: list[tuple[int, int, list]] = []
    cur = None
    for i, c in enumerate(closes):
        if cur is None:
            cur = (1, i, [c])
            continue
        d, start, levels = cur
        if d > 0:
            top = levels[-1]
            if c >= top + box:
                add = int((c - top) // box)
                for k in range(1, add + 1):
                    levels.append(top + box * k)
                continue
            if c <= top - reversal * box:
                new_levels = [top - box]
                add = int((top - box - c) // box)
                for k in range(1, add + 1):
                    new_levels.append(top - box - box * k)
                cols.append((d, start, levels))
                cur = (-1, i, new_levels)
                continue
        else:
            bottom = levels[-1]
            if c <= bottom - box:
                add = int((bottom - c) // box)
                for k in range(1, add + 1):
                    levels.append(bottom - box * k)
                continue
            if c >= bottom + reversal * box:
                new_levels = [bottom + box]
                add = int((c - bottom - box) // box)
                for k in range(1, add + 1):
                    new_levels.append(bottom + box + box * k)
                cols.append((d, start, levels))
                cur = (1, i, new_levels)
                continue
    if cur is not None:
        cols.append((cur[0], cur[1], cur[2]))
    return cols


class PandfBreakoutExpert(Expert):
    """Double/triple top-bottom boxed-price breakout."""
    expert_id = 'pandf_breakout'
    version = 'v1'
    mechanism_family_id = 'boxed_price_breakout'
    behavior_family_id = 'boxed_price_breakout'
    variant_id = 'a'
    # D-044: every implemented variant, losers included. The book card's
    # signal grid enumerates 8 signals (double/triple top/bottom, ascending/
    # descending/spread triple, catapult); the 4 implemented variants plus the
    # 4 dropped (angle-classification triples and the O-013 catapult) are the
    # consumed search universe.
    variants_evaluated = ('a', 'b', 'c', 'd')
    search_universe_size = 8
    requires = ('volatility', 'history')

    def __init__(self, variant_id: str | None = None):
        if variant_id is not None:
            self.variant_id = variant_id

    @staticmethod
    def _signal(cols, variant_id):
        """(direction, anchor_idx, col_bottom, col_top, col_boxes) for the
        signal, or None. A double/triple top requires the current column to be
        an X column whose top exceeds the prior X column top(s); bottoms
        mirror on O columns."""
        if not cols:
            return None
        d_last, start_last, levels_last = cols[-1]
        xs = [c for c in cols if c[0] > 0]
        os_ = [c for c in cols if c[0] < 0]
        if variant_id == 'a':
            if d_last > 0 and len(xs) >= 2 \
                    and levels_last[-1] > xs[-2][2][-1]:
                return ('LONG', start_last, levels_last[0], levels_last[-1],
                        len(levels_last) - 1)
            return None
        if variant_id == 'b':
            if d_last < 0 and len(os_) >= 2 \
                    and levels_last[-1] < os_[-2][2][-1]:
                return ('SHORT', start_last, levels_last[-1], levels_last[0],
                        len(levels_last) - 1)
            return None
        if variant_id == 'c':
            if d_last > 0 and len(xs) >= 3 \
                    and levels_last[-1] > max(xs[-2][2][-1], xs[-3][2][-1]):
                return ('LONG', start_last, levels_last[0], levels_last[-1],
                        len(levels_last) - 1)
            return None
        if d_last < 0 and len(os_) >= 3 \
                and levels_last[-1] < min(os_[-2][2][-1], os_[-3][2][-1]):
            return ('SHORT', start_last, levels_last[-1], levels_last[0],
                    len(levels_last) - 1)
        return None

    def evaluate(self, state: MarketState) -> ExpertEvaluation:
        if self.variant_id not in self.variants_evaluated:
            raise ValueError(
                f'{self.expert_id}: unknown variant {self.variant_id!r} '
                f'(variants_evaluated={list(self.variants_evaluated)})')
        t = state.as_of
        sym = state.universe[0]
        need = [f'{sym}.close', f'{sym}.atr', f'{sym}.history']
        if not self._need(state, need):
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_HABITAT', t)
        f = state.features
        close = float(f[f'{sym}.close'].value)
        atr = f[f'{sym}.atr'].value
        hist_value = f[f'{sym}.history'].value
        if atr is None or atr <= 0 or not isinstance(hist_value, (tuple, list)) \
                or not hist_value or len(hist_value) < MIN_HISTORY_BARS:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_HABITAT', t)
        hist = tuple(hist_value)
        box = float(atr) * BOX_ATR_K
        if box <= 0:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_HABITAT', t)
        closes = [float(b[4]) for b in hist]
        hit = self._signal(_columns(closes, box, REVERSAL_BOXES),
                           self.variant_id)
        if hit is None:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_SETUP', t)
        direction, anchor_idx, col_bottom, col_top, col_boxes = hit
        anchor = hist[anchor_idx][0]
        if direction == 'LONG':
            # Stop below the lowest X in the breakout column; target = the
            # vertical count (Ch15.3): lowest box + boxes * box * reversal.
            stop_price = col_bottom
            target_price = col_bottom + col_boxes * box * REVERSAL_BOXES
            prior_low_ref = col_bottom
            prior_high_ref = None
        else:
            stop_price = col_top
            target_price = col_top - col_boxes * box * REVERSAL_BOXES
            prior_high_ref = col_top
            prior_low_ref = None
        stop_r = (close - stop_price) / float(atr) if direction == 'LONG' \
            else (stop_price - close) / float(atr)
        target_r = (target_price - close) / float(atr) if direction == 'LONG' \
            else (close - target_price) / float(atr)
        if stop_r <= 0 or target_r <= 0:
            return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                    'NOT_APPLICABLE', 'NO_SETUP', t)
        geometry = {
            'entry': 'NEXT_BAR_CLOSE', 'target_r': target_r, 'stop_r': stop_r,
            'expiry_bars': 8, 'atr_ref': float(atr), 'variant': self.variant_id,
            'reversal': REVERSAL_BOXES,
        }
        if prior_low_ref is not None:
            geometry['prior_low_ref'] = prior_low_ref
        if prior_high_ref is not None:
            geometry['prior_high_ref'] = prior_high_ref
        draft = CandidateDraft(
            expert_id=self.expert_id, expert_version=self.version,
            instrument=sym, direction=direction,
            setup_fingerprint=f'{sym}:{self.variant_id}:{direction}:{close:.6f}',
            risk_geometry=geometry, birth_time=t,
            setup_anchor_event_id=anchor)
        return ExpertEvaluation(self.expert_id, self.version, state.state_id,
                                'APPLICABLE', 'CANDIDATE', t, draft)

    def still_valid(self, state: MarketState, draft: CandidateDraft) -> bool:
        """The thesis is "the boxed breakout is in force": a long must stay
        above the frozen breakout reference (the lowest X in the breakout
        column — the level a close below would retrace the whole column), a
        short below the highest O. A close through the reference says the
        breakout failed. Fails open when the close is unobservable."""
        sym = draft.instrument
        f = state.features
        close = f.get(f'{sym}.close')
        if close is None or close.value is None:
            return True
        close = float(close.value)
        if draft.direction == 'LONG':
            ref = draft.risk_geometry.get('prior_low_ref')
            if ref is None:
                return True
            return close > float(ref)
        ref = draft.risk_geometry.get('prior_high_ref')
        if ref is None:
            return True
        return close < float(ref)

File: v8-core/src/experts/pandf_breakout.rs
//! pandf_breakout: evaluate() port target (issue #95) — mirror src/v8/experts/pandf_breakout.py
//! bit-for-bit (PARITY_AND_IDENTITY_SPEC §3; COMPUTE_CORE_SPEC §8 S4).
//!
//! Point-and-figure boxed-price breakout (variants a/b/c/d). The P&F transform
//! is computed inside the expert from the history close window: columns of X
//! rising / O falling, box = 1.0 * ATR at detection (LOCKED orchestrator
//! directive), reversal = 3 boxes. The setup anchor is the column-START index
//! of the current (breakout) column (`hist[anchor_idx][0]`), NOT the D-026
//! find_setup_anchor run-start — the Python source anchors on the column.
//!
//! The parity harness constructs `PandfBreakoutExpert()` with its default
//! variant_id 'a', and the port fn signature carries no variant parameter, so
//! the declared default variant 'a' is used here; the full `_signal` machinery
//! (all four variants) is mirrored below for fidelity.

use crate::experts::base::*;
use crate::simulator::Draft;

pub const PORTED: bool = true;
pub const VERSION: &str = "v1";
pub const REQUIRES: &[&str] = &["volatility", "history"];
// Declared risk geometry (EXPERT_PROTOCOL §1: risk geometry is "Predeclared
// entry, stop, target, timeout and sizing inputs"; SIMULATION_TRUTH_SPEC D-028:
// R is a declared price distance). Fixed values are declared here, never
// re-literalized inside evaluate(); a structural target/stop is computed at
// the call site and overrides the key.
pub const EXPIRY_BARS: i64 = 8;

// Declared, LOCKED box filter (orchestrator directive): box = 1.0 * ATR at
// detection, reversal = 3 boxes.
const BOX_ATR_K: f64 = 1.0;
const REVERSAL_BOXES: i64 = 3;
// Minimum history for a meaningful column structure.
const MIN_HISTORY_BARS: usize = 20;

/// One P&F column: direction (+1 = X rising, -1 = O falling), index of the
/// first close that contributed, and box prices ordered from the column's
/// origin to its extreme (rising for X, falling for O).
struct Column {
    d: i64,
    start: usize,
    levels: Vec<f64>,
}

/// A matched breakout signal (mirror of the Python `_signal` return tuple
/// (direction, anchor_idx, col_bottom, col_top, col_boxes)).
struct Signal {
    direction: &'static str,
    anchor_idx: usize,
    col_bottom: f64,
    col_top: f64,
    col_boxes: usize,
}

/// Deterministic close-based P&F columns (mirror of `_columns`).
fn columns(closes: &[f64], box_: f64, reversal: i64) -> Vec<Column> {
    let mut cols: Vec<Column> = Vec::new();
    let mut cur: Option<Column> = None;
    for (i, &c) in closes.iter().enumerate() {
        if cur.is_none() {
            cur = Some(Column { d: 1, start: i, levels: vec![c] });
            continue;
        }
        let col = cur.as_mut().unwrap();
        if col.d > 0 {
            let top = *col.levels.last().unwrap();
            if c >= top + box_ {
                let add = ((c - top) / box_) as i64;
                for k in 1..=add {
                    col.levels.push(top + box_ * (k as f64));
                }
                continue;
            }
            if c <= top - (reversal as f64) * box_ {
                let mut new_levels = vec![top - box_];
                let add = ((top - box_ - c) / box_) as i64;
                for k in 1..=add {
                    new_levels.push(top - box_ - box_ * (k as f64));
                }
                cols.push(std::mem::replace(
                    col,
                    Column { d: 0, start: 0, levels: Vec::new() },
                ));
                cur = Some(Column { d: -1, start: i, levels: new_levels });
                continue;
            }
        } else {
            let bottom = *col.levels.last().unwrap();
            if c <= bottom - box_ {
                let add = ((bottom - c) / box_) as i64;
                for k in 1..=add {
                    col.levels.push(bottom - box_ * (k as f64));
                }
                continue;
            }
            if c >= bottom + (reversal as f64) * box_ {
                let mut new_levels = vec![bottom + box_];
                let add = ((c - bottom - box_) / box_) as i64;
                for k in 1..=add {
                    new_levels.push(bottom + box_ + box_ * (k as f64));
                }
                cols.push(std::mem::replace(
                    col,
                    Column { d: 0, start: 0, levels: Vec::new() },
                ));
                cur = Some(Column { d: 1, start: i, levels: new_levels });
                continue;
            }
        }
    }
    if let Some(c) = cur {
        cols.push(c);
    }
    cols
}

/// (direction, anchor_idx, col_bottom, col_top, col_boxes) for the signal, or
/// None. A double/triple top requires the current column to be an X column
/// whose top exceeds the prior X column top(s); bottoms mirror on O columns.
fn signal(cols: &[Column], variant_id: &str) -> Option<Signal> {
    if cols.is_empty() {
        return None;
    }
    let last = cols.last().unwrap();
    let d_last = last.d;
    let start_last = last.start;
    let levels_last = &last.levels;
    let xs: Vec<&Column> = cols.iter().filter(|c| c.d > 0).collect();
    let os_: Vec<&Column> = cols.iter().filter(|c| c.d < 0).collect();
    match variant_id {
        "a" => {
            if d_last > 0
                && xs.len() >= 2
                && levels_last[levels_last.len() - 1]
                    > xs[xs.len() - 2].levels[xs[xs.len() - 2].levels.len() - 1]
            {
                return Some(Signal {
                    direction: "LONG",
                    anchor_idx: start_last,
                    col_bottom: levels_last[0],
                    col_top: levels_last[levels_last.len() - 1],
                    col_boxes: levels_last.len() - 1,
                });
            }
            None
        }
        "b" => {
            if d_last < 0
                && os_.len() >= 2
                && levels_last[levels_last.len() - 1]
                    < os_[os_.len() - 2].levels[os_[os_.len() - 2].levels.len() - 1]
            {
                return Some(Signal {
                    direction: "SHORT",
                    anchor_idx: start_last,
                    col_bottom: levels_last[levels_last.len() - 1],
                    col_top: levels_last[0],
                    col_boxes: levels_last.len() - 1,
                });
            }
            None
        }
        "c" => {
            if d_last > 0
                && xs.len() >= 3
                && levels_last[levels_last.len() - 1]
                    > xs[xs.len() - 2].levels[xs[xs.len() - 2].levels.len() - 1]
                        .max(xs[xs.len() - 3].levels[xs[xs.len() - 3].levels.len() - 1])
            {
                return Some(Signal {
                    direction: "LONG",
                    anchor_idx: start_last,
                    col_bottom: levels_last[0],
                    col_top: levels_last[levels_last.len() - 1],
                    col_boxes: levels_last.len() - 1,
                });
            }
            None
        }
        _ => {
            if d_last < 0
                && os_.len() >= 3
                && levels_last[levels_last.len() - 1]
                    < os_[os_.len() - 2].levels[os_[os_.len() - 2].levels.len() - 1]
                        .min(os_[os_.len() - 3].levels[os_[os_.len() - 3].levels.len() - 1])
            {
                return Some(Signal {
                    direction: "SHORT",
                    anchor_idx: start_last,
                    col_bottom: levels_last[levels_last.len() - 1],
                    col_top: levels_last[0],
                    col_boxes: levels_last.len() - 1,
                });
            }
            None
        }
    }
}

pub fn pandf_breakout(fm: &FeatMap, expert_id: &str, version: &str) -> ExpertEval {
    // The parity harness instantiates PandfBreakoutExpert() with its default
    // variant_id 'a'; the port fn signature carries no variant, so the default
    // is used (mirror of the Python __init__ default).
    let variant_id = "a";
    let sym = fm.symbol;
    let close = match fm.value("close") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    let atr = match fm.value("atr") {
        Some(v) => v,
        None => return no_habitat(expert_id, version, fm.as_of),
    };
    // Python: atr None or atr <= 0 or empty history / len < MIN_HISTORY_BARS.
    if atr <= 0.0 || fm.history.len() < MIN_HISTORY_BARS {
        return no_habitat(expert_id, version, fm.as_of);
    }
    let box_ = atr * BOX_ATR_K;
    if box_ <= 0.0 {
        return no_habitat(expert_id, version, fm.as_of);
    }
    let closes: Vec<f64> = fm.history.iter().map(|b| b.close).collect();
    let hit = match signal(&columns(&closes, box_, REVERSAL_BOXES), variant_id) {
        Some(h) => h,
        None => return no_setup(expert_id, version, fm.as_of),
    };
    // The setup anchor is the event id of the bar at the breakout column's
    // start index (hist[anchor_idx][0]) — NOT the D-026 find_setup_anchor.
    let anchor = fm.history[hit.anchor_idx].event_id.clone();
    let (stop_price, target_price, prior_low_ref, prior_high_ref) = if hit.direction == "LONG" {
        let stop_price = hit.col_bottom;
        let target_price =
            hit.col_bottom + (hit.col_boxes as f64) * box_ * (REVERSAL_BOXES as f64);
        (stop_price, target_price, Some(hit.col_bottom), None)
    } else {
        let stop_price = hit.col_top;
        let target_price = hit.col_top - (hit.col_boxes as f64) * box_ * (REVERSAL_BOXES as f64);
        (stop_price, target_price, None, Some(hit.col_top))
    };
    let stop_r = if hit.direction == "LONG" {
        (close - stop_price) / atr
    } else {
        (stop_price - close) / atr
    };
    let target_r = if hit.direction == "LONG" {
        (target_price - close) / atr
    } else {
        (close - target_price) / atr
    };
    if stop_r <= 0.0 || target_r <= 0.0 {
        return no_setup(expert_id, version, fm.as_of);
    }
    let mut entries = vec![
        ("entry", serde_json::json!("NEXT_BAR_CLOSE")),
        ("target_r", serde_json::json!(target_r)),
        ("stop_r", serde_json::json!(stop_r)),
        ("expiry_bars", serde_json::json!(EXPIRY_BARS)),
        ("atr_ref", serde_json::json!(atr)),
        ("variant", serde_json::json!(variant_id)),
        ("reversal", serde_json::json!(REVERSAL_BOXES)),
    ];
    if let Some(v) = prior_low_ref {
        entries.push(("prior_low_ref", serde_json::json!(v)));
    }
    if let Some(v) = prior_high_ref {
        entries.push(("prior_high_ref", serde_json::json!(v)));
    }
    let draft = Draft {
        direction: hit.direction.to_string(),
        birth_time: fm.as_of,
        risk_geometry: geom(entries),
    };
    let fingerprint = format!("{sym}:{variant_id}:{}:{:.6}", hit.direction, close);
    candidate(expert_id, version, fm.as_of, draft, anchor, fingerprint)
}
